src/dependency.cr

```diff
@@ -7,8 +7,105 @@ module Shards
+
+    def initialize(@name : String, @resolver : Resolver, @requirement : Requirement = Any, @resolver_key : String? = nil, @source : String? = nil)
+    end
+
+    # Parse a dependency from a CLI argument
+    def self.from_cli(value : String) : Dependency
+      parts = parts_from_cli(value)
 
-    def initialize(@name : String, @resolver : Resolver, @requirement : Requirement = Any)
+      # We need to check the actual shard name to create a dependency.
+      # This requires getting the actual spec file from some matching version.
+      resolver = Resolver.find_resolver(parts[:resolver_key], "unknown", parts[:source])
+      version = resolver.versions_for(parts[:requirement]).first || raise Shards::Error.new("No versions found for dependency: #{value}")
+      spec = resolver.spec(version)
+      name = spec.name || raise Shards::Error.new("No name found for dependency: #{value}")
+
+      Dependency.new(name, resolver, parts[:requirement], parts[:resolver_key], parts[:source])
+    end
+
+    # :nodoc:
+    #
+    # Parse the dependency from a CLI argument
+    # and return the parts needed to create the proper dependency.
+    #
+    # Split to allow better unit testing.
+    def self.parts_from_cli(value : String) : {resolver_key: String, source: String, requirement: Requirement}
+      resolver_key = nil
+      source = ""
+      requirement = Any
+
+      if File.directory?(value)
+        resolver_key = "path"
+        source = value
+      end
+
+      if value.starts_with?("https://github.com")
+        resolver_key = "github"
+        uri = URI.parse(value)
+        source = uri.path[1..-1] # drop first "/""
+
+        components = source.split("/")
+        case components[2]?
+        when "commit"
+          source = "#{components[0]}/#{components[1]}"
+          requirement = GitCommitRef.new(components[3])
+        when "tree"
+          source = "#{components[0]}/#{components[1]}"
+          requirement = if components[3].starts_with?("v")
+                          GitTagRef.new(components[3])
+                        else
+                          GitBranchRef.new(components[3..-1].join("/"))
+                        end
+        end
+      end
+
+      if value.starts_with?("https://gitlab.com")
+        resolver_key = "gitlab"
+        uri = URI.parse(value)
+        source = uri.path[1..-1] # drop first "/""
+      end
+
+      if value.starts_with?("https://bitbucket.com")
+        resolver_key = "bitbucket"
+        uri = URI.parse(value)
+        source = uri.path[1..-1] # drop first "/""
+      end
+
+      if value.starts_with?("git://")
+        resolver_key = "git"
+        source = value
+      end
+
+      unless resolver_key
+        Resolver.resolver_keys.each do |key|
+          key_schema = "#{key}:"
+          if value.starts_with?(key_schema)
+            resolver_key = key
+            source = value.sub(key_schema, "")
+
+            # narrow down requirement
+            if source.includes?("@")
```

Comment:
What about a SSH URL? `git@host.com/user/repo.git` ?

Reply:
Is `git@` always the prefix?

Reply:
The username can be anything. `git` is just a common name used with key-based authentication.

Reply:
The fully-qualified URI would be `git:user@host/path`.
We can use the `git@` prefix as a heuristic for the git resolver, though.